bluesky_widgets/qt/threading.py

```diff
@@ -0,0 +1,625 @@
+        A function to call in another thread
+    *args
+        will be passed to the function
+    **kwargs
+        will be passed to the function
+
+    Raises
+    ------
+    TypeError
+        If ``func`` is a generator function and not a regular function.
+    """
+
+    def __init__(self, func: Callable, *args, **kwargs):
+        if inspect.isgeneratorfunction(func):
+            raise TypeError(
+                f"Generator function {func} cannot be used with "
+                "FunctionWorker, use GeneratorWorker instead"
+            )
+        super().__init__()
+
+        self._func = func
+        self._args = args
+        self._kwargs = kwargs
+
+    def work(self):
+        return self._func(*self._args, **self._kwargs)
+
+
+class GeneratorWorkerSignals(WorkerBaseSignals):
+
+    yielded = Signal(object)  # emitted with yielded values (if generator used)
+    paused = Signal()  # emitted when a running job has successfully paused
+    resumed = Signal()  # emitted when a paused job has successfully resumed
+    aborted = Signal()  # emitted when a running job is successfully aborted
+
+
+class GeneratorWorker(WorkerBase):
+    """QRunnable with signals that wraps a long-running generator.
+
+    Provides a convenient way to run a generator function in another thread,
+    while allowing 2-way communication between threads, using plain-python
+    generator syntax in the original function.
+
+    Parameters
+    ----------
+    func : callable
+        The function being run in another thread.  May be a generator function.
+    SignalsClass : type, optional
+        A QObject subclass that contains signals, by default
+        GeneratorWorkerSignals
+    *args
+        Will be passed to func on instantiation
+    **kwargs
+        Will be passed to func on instantiation
+    """
+
+    def __init__(
+        self,
+        func: Callable,
+        *args,
+        SignalsClass: Type[QObject] = GeneratorWorkerSignals,
+        **kwargs,
+    ):
+        if not inspect.isgeneratorfunction(func):
+            raise TypeError(
+                f"Regular function {func} cannot be used with "
+                "GeneratorWorker, use FunctionWorker instead"
+            )
+        super().__init__(SignalsClass=SignalsClass)
+
+        self._gen = func(*args, **kwargs)
+        self._incoming_value = None
+        self._pause_requested = False
+        self._resume_requested = False
+        self._paused = False
+        # polling interval: ONLY relevant if the user paused a running worker
+        self._pause_interval = 0.01
+
+    def work(self) -> None:
+        """Core event loop that calls the original function.
+
+        Enters a continual loop, yielding and returning from the original
+        function.  Checks for various events (quit, pause, resume, etc...).
+        (To clarify: we are creating a rudimentary event loop here because
+        there IS NO Qt event loop running in the other thread to hook into)
+        """
```

Comment:
I am not entirely sure what the main reason for this comment. Maybe I am not understanding what it means when the comment say there is NO Qt event loop in the other thread. You can always construct things in a way you can create a QEventLoop at the appropriate level to hook it.